File: solution/c#/Day22/Day22/Diamond.cs
using LanguageExt;
using static System.Environment;
using static System.Linq.Enumerable;
using static System.String;

namespace Day22
{
    public static class Diamond
    {
        private const char Start = 'A';
        public static Option<string> ToDiamond(this char c) => Print(c);

        private static Option<string> Print(char c)
            => IsValidCharacter(c)
                ? FullDiamondSafely(c)
                : Option<string>.None;

        private static bool IsValidCharacter(char endCharacter)
            => endCharacter is >= Start and <= 'Z';

        private static string FullDiamondSafely(char endCharacter)
            => Join(NewLine, GenerateDiamond(endCharacter));

        private static Seq<string> GenerateDiamond(char endCharacter)
            => GenerateHalfDiamond(endCharacter)
                .Apply(halfDiamond => halfDiamond.Concat(
                    halfDiamond.Reverse().Tail()
                ));

        private static Seq<string> GenerateHalfDiamond(char endCharacter)
            => Range(Start, endCharacter - Start + 1).ToSeq()
                .Map(i => (char) i)
                .Map(c => ToLine(c, endCharacter));

        private static string ToLine(char character, char endCharacter)
            => Outer(character, endCharacter)
                .Apply(@out => @out + character + Inner(character) + @out);

        private static string Outer(char character, char endCharacter)
            => GenerateEmptyCharacters(endCharacter - character);

        private static string Inner(char character)
            => character != Start
                ? GenerateEmptyCharacters(NumberOfEmptyCharactersFor(character)) + character
                : "";

        private static int NumberOfEmptyCharactersFor(char character)
            => (character - Start) * 2 - 1;

        private static string GenerateEmptyCharacters(int count)
            => Range(0, count).ToSeq()
                .Fold("", (acc, _) => acc + " ");
    }

    public static class FunctionalExtensions
    {
        public static T Apply<T>(this T obj, Func<T, T> action) => action(obj);
    }
}

File: solution/ts/day22/src/diamond.ts
import {none, Option, some} from "fp-ts/Option";

export class Diamond {
    private static readonly START = 'A';

    public static print = (endCharacter: string): Option<string> =>
        this.isValidCharacter(endCharacter)
            ? some(this.fullDiamondSafely(endCharacter))
            : none;

    private static fullDiamondSafely(endCharacter: string): string {
        const halfDiamond = this.generateHalfDiamond(endCharacter);
        const fullDiamond = halfDiamond.concat(halfDiamond.slice(0, -1).reverse());

        return fullDiamond.join('\n');
    }

    private static generateHalfDiamond = (endCharacter: string): string[] =>
        Array.from({length: endCharacter.charCodeAt(0) - this.START.charCodeAt(0) + 1}, (_, index) => {
            const currentChar = String.fromCharCode(this.START.charCodeAt(0) + index);
            return this.toLine(currentChar, endCharacter);
        });

    private static toLine(character: string, endCharacter: string): string {
        const outerSpaces = this.generateSpaces(endCharacter.charCodeAt(0) - character.charCodeAt(0));
        const innerSpaces = character === this.START ? '' : this.generateSpaces((character.charCodeAt(0) - this.START.charCodeAt(0)) * 2 - 1) + character;
        return outerSpaces + character + innerSpaces + outerSpaces;
    }

    private static generateSpaces = (count: number): string => ' '.repeat(count);

    private static isValidCharacter = (endCharacter: string): boolean => endCharacter >= this.START && endCharacter <= 'Z';
}